README.md
# MOULINETTE
MOULINETTE:  SRA Sequence Search Tool (Srasst) 


Pipeline to report OTU ocurrence of an organism

![alt text](https://github.com/ybadis/moulinette/raw/master/images/pipeline.png "Moulinette pipeline")

## How does it work?

A query in fasta (or Multi query fasta file) is run against a list of SRA runs. For each run, the query is blasted (blastn-vdb; SRA-toolkit 2.8.0, http://www.ncbi.nlm.nih.gov/Traces/sra) on the run. Based on blast results, reads matching any query above the selected identity, e-value, and alignment length thresholds are retrieved in fastq (fastq-dump, SRA-toolkit) in separate folders corresponding to each SRA run. A final result fastq file and count file is generated and each dataset is cleared after processing. 

This data reduction approach allows to screen for large amount of data with no storage concerns. In many instances, the MOULINETTE can retrieve the GPS coordinates of any given SRA Dataset, provided that those metadata are correctly assigned by submitters (e.g Ocean Sampling Day dataset). The pipeline accepts single and paired reads datasets, and those are stored in separate files in fastq format. The current version sends only paired-end reads (pooled reads from all screened datasets) to the USEARCH OTU calling pipeline, using a filtering parameter of 1.0 (expected error), an OTU radius of 3 (clustering based on 97% Identity), and excluding singletons for OTU calling. A remote blast is conducted to retrieve the best match of each defined OTU on the NCBI N/R database. The OTUs defined are finally gathered in a blast database, and all reads (paired and single end) are blasted against OTUs to assign them to their best matching OTU with 97% threshold, thus allowing the definition of biogeographical patterns based on GPS coordinate retrieval.


### (a) - Selection of Input SRA Runs

The pipeline can handle broad range of datasets, selected with or without a priori. Because MOULINETTE analyses one SRA run after the other, and keeps only reads satisfying the user-criteria, the input list of SRA Runs can contain irrelevant data (e.g A full SRA Bioproject containing both eukaryotic 18S and bacterial 16s barcoding data while the user is only interested in 18S data) as it will not yield any result and will not be integrated in any kind of normalization, however this will significantly increase the length of a MOULINETTE run. In this paper, it is the case for the selection of 19000 SRA Freshwater datasets: to our knowledge, and as of January 2017, this selection covered all possible 18S freshwater barcoding SRA runs, but also contained 16S data and whole metagenome sequencing. In such instances, including irrelevant data and increasing the length of the MOULINETTE run is quicker than manually sorting only relevant datasets. The selection of various runs from different studies is not common, and has advantages and limitations that are discussed below. Merging MOULINETTE Runs for large input lists: The current version of the pipeline (V11.3) does not handle multithreading and processes only one run at a time. Users can split a list of datasets in subsets of 1000-2000 SRA runs, and launch several MOULINETTE runs in parallel. Independent output result folders can be merged with the MOULINETTE-MergeRunsV4 script, available upon request.

 

### (b) - Selection of Input query sequences

The main current purpose of the MOULINETTE pipeline is to screen metabarcoding data, to detect which SRA raw dataset might contain reads matching an organism of interest. Although many other uses are envisaged (e.g SNP discovery, gene reconstruction in environmental whole DNA sequencing), the current design of the pipeline has been focussed on metabarcoding strategies. In that sense, queries should be limited to classic markers like 18S, 16S, COI and others, and we encourage users to trim queries to suit their needs. For example, the v4 and v9 regions of the 18s marker of an organism of interest can be considered as two separate queries, this would reduce potential false positive originating from reads matching non variable regions of the markers. Alternatively, retrieved reads can be sorted based on the position of the blast alignment (e.g. locating reads matching v4 or v9 regions for a particular full length sequence). Both approaches yield similar results.

### (c) - Stringency Parameters – E-value

We provided the user with the possibility to choose an e-value threshold for retrieval of meaningful matching reads. This ensures flexibility of the tool for future developments. However, selecting an appropriate e-value threshold is tedious and highly dependent on the queries used (length), and likely to be irrelevant with multi-fasta queries. Additionally, because each SRA dataset is considered as an independent blast database through vdb-blast, e-value scores could vary for the same read between two different SRA RUNs (e.g two SRA runs can have very different sizes, in some instances few thousand reads, in other instances millions of reads). We recommend users to set the e-value threshold to a relaxed stringency (e.g 1e-40) and rely on the two parameters discussed below for retrieval of meaningful information. For reference, in the SRA run ERR867676, containing 115662 reads, a read matching a given query at 100% identity typically has a blast e-value below 1e-100. 

### (d) - Stringency Parameters - %ID and READCOVER

The percentage of identity and the “READCOVER” parameter ($3 and $5 in the Usage section above, respectively) are the main parameters to take in account for accurate retrieval of meaningful hits. The READCOVER parameter is crucial: For a value of 0.8, the pipeline retains a read if the length of the aligned region represents more than 80% of the read length. While many reads could match a query at 100% identity on only few nucleotides (false positives), this filtering is performed upon processing of each read blast result, and ensures that only meaningful alignments retained despite the low stringency e-value mentioned above. Stringent blast parameters (e.g 100% Identity) will retrieve the biogeography of a target organism (i.e only one SRA-OTU is retrieved), while relaxed parameters (e.g 97% identity) will allow to explore the associated OTU diversity and Biogeography, although this can vary depending on paired and single read layouts as discussed below.

### (e) - Screening of Paired-end data Vs Single Read Data

The current design of the pipeline uses vdb-blast to harvest reads matching a given query. In the case of paired-end layout, Vdb-blast considers forward and reverse reads separately and provides two blast results for a read pair. For the 18S V4 region amplified by classical Eukaryotic primers, the most variable region is not centred, See figure below. 

![alt text](https://raw.githubusercontent.com/ybadis/moulinette/master/images/18Sv4region.png "18S V4 region amplified by classical Eukaryotic primers,")

 
In the example shown above, the forward read contains most of the variable region, and will be much more sensitive at discriminating species. The reverse read spans a less variable region and will tend to be less specific. In term of identity scores a barcode (final merged read) 98.5% identical to our target organism is in fact made of two reads with 97.3% and 99.6% identity. This notion is important because the MOULINETTE will harvest both F+R reads of a read pair, even if only one read satisfies the user defined stringency criteria. A user threshold of 97% can thus harvest a Reverse read at 97.1% identity that will give a barcode (merged read) below this threshold, for example approx. 95% Identity to our target organism. This can be useful when only few known sequences cluster close from a target organism, for example in the major but distantly related Olpidiopsis clades of this study, where they share an average 90-92% identity on their V4 region. Such phenomenon is not problematic for the cautious user, as the limited number of final OTUs allows for manual curation. The “De-Novo Targeted OTU” strategy of the MOULINETTE typically give a small number of OTUs (e.g 1 to 30 Vs over 1000 for classic metabarcoding) that are reasonably easy to check and process compared with regular OTU strategies. Note that this does not apply to single read layouts, as they represent the equivalent of the merged read discussed above. In conclusion, the %Id stringency parameter does not have the same meaning for paired vs single read layouts. Future implementations of the pipeline will address this potential limitation through the use of the recently released Magic-BLAST, that can generate blast results taking into account simultaneously the two reads of a pair.

### (f) - OTU Calling

Only paired-end reads are used to call OTUs via the Usearch pipeline, but both paired-end and single-end reads are mapped to OTUs. Future developments of the MOULINETTE Pipeline, will aim at integrating both read types prior to OTU calling. The clustering threshold used by Usearch can be modified in the script (otu radius parameter), and should be adapted to the species delimitation offered by a region of interest. In this study, we chose the 97% clustering threshold (out radius 3) mostly as this standard is use throughout many metabarcoding campaigns, and to be coherent with the OTUs generated by the independent Australian Marine Microbe Project and the P.umbilicalis metabarcoding generated at JGI.

### (g) - Read Mapping

The MOULINETTE pipeline uses regular blast based scores to assign reads to their best matching OTU. Although not using the otu_tab command of the Usearch pipeline, our approach is consistent with the guidelines and discussion provided in the Usearch website (http://www.drive5.com/usearch/manual/mapreadstootus.html). Guidelines of the Usearch website suggest to map merged unfiltered reads to account for a larger fraction of the reads (http://www.drive5.com/usearch/manual/cmd_otutab.html), however considering the low number of read per organism for rare obligate parasites, we chose to also map unmerged reads provided that their reciprocal best blast hit in N/R is an Olpidiopsis sequence. This concerns 135 reads out of the 1283 final read pool, we manually checked that no SRA Runs were retained in the final Biogeography dataset based on the sole presence of such unmerged reads.

### (h) - OTU Occurrence Vs Abundance

The MOULINETTE pipeline is aimed at reporting the occurrence of an organism, but not its abundance. Indeed, as OTU calling is performed on a pool of reads from a whole diversity of datasets (potentially different sequencing technologies, DNA extraction protocols, years) complex normalization strategies would be needed to compare abundance between unrelated SRA Runs. As a consequence, the current version of the pipeline and our definition of OTUs is not appropriate to infer diversity metrics.

### (i) - Cross talk effects

As pooled SRA datasets of a large MOULINETTE run could have been generated on different machines - e.g pooling different SRA bioprojects or different years -  cross-talk effects (contamination of read, assignment of reads to the wrong sample during demultiplexing etc..) are intrinsically less likely. However, the cross-talk effects of related SRA runs (e.g. from the same SRA Bioprojects) in a whole Moulinette run could still give rise to spurious low count reads as discussed in http://www.drive5.com/usearch/manual/otu_count_interpret.html

### (j) - Spurious low count OTUs and Limitations of Biogeography


CHANGELOG.md
# Changelog

All changes to this software will be documented in this file

## v11.5 

- updated GPS extraction, now accepting multiple formats using the dedicated xtract xml parser utility
- only 100 first reads used for computing read length

##  v11.3 

- maxtargetseq set to 1 in the final NR blast to have only the best hit retrieved and facilitate sorting which OTU are of interest. Users will run global remote blasts themselves using maxtargetseq 10 to get the diversity of best matches
- clustering OTU with Minsize 2 --> Excluding singletons
- took out many * wilcards in the cleanup stage so that cp and mv are more precise, allowing running several instances of the moulinette in the same working directory

## v11.2 

- added "--skip-technical to all fastq-dump commands to keep only Biological Reads"
- added "n" to the sort command in the definition of ALNLGT lgt

## v11

- test each retrieved read to see if paired or not to send to two global file, one for single reads, one for paired reads, only the paired reads will be sent through OTU calling
- added a command to save all stdout to final logfile for future debugging


## v10.1 

- add "sort -u" before each fastq-dump to avoid extractin same read several times

## v10.0 

- Implementing final reciprocal blast on of retrieved reads VS usearch Otus - see line 324, reciprocal blast will be made on R1 and R2 reads, but also on merged reads produced by usearch before the filtration step
- also includes remote blast of final OTU list
- I also took outh the relabel option of usearch for enhanced trancability

## v9.2 --> FDB MOD

- ALNLGT variable definition changed;solving the issue of wrapped vs unwrapped read data by grepping "lenght=" of reads using fastq-dump instead of lenght of lines
- changed filenames of executables, removing the version IDs find-replace, cach-mgr to ./cache-mgr to use the working directory

## v9 

-Retrieving GPS before fastqdump to generate per-read global result file
## v8
- GPS- Extraction at each Run and not the end of run. -- v7: addition of 
- SraSST-GPS-Runextract module and Osearch OTU calling to SraSST-v6.6)





SraSST.sh
#!/bin/bash

##	Release v11.7 03-12-19
##	Contact yacinebadis@sams.ac.uk


################## DESCRIPTION - USAGE - OUTPUT
## Description: A Query $1 is run against  A list of SRA Runs $2. For each run, the query is vdb-blasted on the Run and Reads above the selected 
## alignment length threshholds are retrieved in fastq and converted to fasta in separate folders corresponding to each SRA Run.
## A Final result fastq file and count file is generated. Runs with no hits should not be saved. 

## Requirements: sratoolkit (tested with sratoolkit.2.8.0-centos_linux64); entrezdirect (tested with Sept 2016 package); usearch (tested with usearch v9.1.13_i86linux32)

##### Important  Note: Stream of data does not mean data are not in cache. Up to a certain extent, keeping cache data allows 
##### extra fast retrieval on RUNs that were already queried --> interesting when automated with Many queries following intial scan of a bif runlist file.
##### Data are stored in the ncbi/cache folder,  attention should be paid tokeep the cache folder under control. Per default, cache is emptied after each
##### run, this can be modified by switching cache option -c (clear) to -r (read) near line 264. 

### Usage:  bash ./SraSST-v7.sh $1 $2 $3 $4 $5 $6   
### Usage example for metabarcoding runs: bash ./SraSST-v7.sh query.fasta runlist.txt 98.5 100 0.8 1e-100
### Query can be a multifasta

###OUTPUT will contain In One global SraSST run folder:
### - an *input folder containing copies of original input files 
### - a log file summarizing results and parameters used for the run (*.log)
### - a global count table summarizing the number of read pairs retrieved for each positive run (*counts.tab)
### - a global fastq file pooling all retrieved reads (*All-retrieved-reads-paired.fastq)
### - a global table containg GPS coordinates of every positive sra run (If such metadata were provided by submitters to sra)
### - a folder containg the whole usearch outputs generated (*-usearch folder, the final otu file has the *otus.fa suffix)
### - a folder for each positive run containing (i) blast results (ii) selected ids (iii) a fastq file for each extracted read pair

##Enter Debug mode (deactivated here)
#set -x

PATH="/opt/sratoolkit/2.9.6/bin:$PATH"
PATH="/opt/edirect:$PATH"
#PATH="/opt/usearch/9.1.13:$PATH"
PATH="/opt/usearch/9.2.64:$PATH"
#PATH="/opt/usearch/11.0.667:$PATH"
PATH="/opt/ncbi-blast/2.9.0/bin:$PATH"

NTHREADS=16

usage() { 
	echo "Usage: $0 [-q <fasta>] [-l <list>] [-t <threshold>] [-n <ntarget>] [-p <percentage>] [-e <evalue>] [-o <dir>]" 1>&2
	echo "fasta: file e.g. Ectrogella or Olpidiopsis"
	echo "list: name of input list of sra RUN in txt format (e.g runselector output on ncbi, e.g ERR867907 and many others)"
	echo "threshold: blast identity threshold"
	echo "ntarget: max number of target sequences in blast parameter"
	echo "percentage: percentage read cover for blast algntlgt filter expressed in decimal e.g 0.8"
	echo "evalue: Blast evalue parameter"
	echo "dir: output directory"
	echo "Usage example for metabarcoding:"
        echo "./SraSST.sh -q query.fa -l runlist.txt -t 98.5 -n 100 -p 0.8 -e 1e-100 -o out"
	exit 1
}

while getopts ":h:q:l:t:n:p:e:o:" o; do
    case "${o}" in
	q) QUERY=${OPTARG};; # fasta file e.g. Ectrogella or Olpidiopsis
        l) LIST=${OPTARG} ;; # name of input list of sra RUN in txt format (e.g runselector output on ncbi, e.g ERR867907)
        t) PERCID=${OPTARG} ;; # blast identity threshold
        n) MAXTARGET=${OPTARG} ;; # max number of target sequences in blast parameter
        p) READCOVER=${OPTARG} ;; # percentage read cover for blast algntlgt filter expressed in decimal e.g 0.8
        e) EVALUE=${OPTARG} ;; #  Blast evalue parameter
        o) OUTPUTDIR=${OPTARG} ; mkdir -p $OUTPUTDIR;; #  Outputdir
	\?) echo "Invalid option: -$OPTARG" >&2;exit ;;
        *) usage ;;
    esac
done
# shift so that $@, $1, etc. refer to the non-option arguments
shift $((OPTIND-1))



#----------------------------------INPUT VALIDATION----------------------

RUNCOUNT=$(cat "$LIST" | wc -l)
QUERYNB=$(cat "$QUERY" | grep ">" | wc -l)

echo QUERY is "$QUERY"  fasta queryfile
echo LIST is "$LIST"  name of input list of sra RUN 
echo PERCID "$PERCID"  blast identity threshold
echo MAXTARGET is "$MAXTARGET" "max number of target sequences in blast parameter"
echo READCOVER is "$READCOVER"  "percentage read cover for blast algntlgt filtering"
echo EVALUE is "$EVALUE"  "Blast evalue parameter"
echo OUTPUT Folder is "$OUTPUTDIR"
echo ""
echo ""
echo Launching SraSST RUN with "$RUNCOUNT" sra runs and $QUERYNB fasta query sequence-s
echo In Our hands, a single metagenomic run can typically take 0 to 15 min to scan with 9 queries
echo THIS MIGHT BE LONG ...
echo ""
echo ""

while true;
do
	echo "Please check your input variables."
	echo ""
	echo ""
	echo -n " Do you wish to proceed? Please confirm (y or n) :"
	read CONFIRM
	case $CONFIRM in
		y|Y|YES|yes|Yes) break ;;
		n|N|no|NO|No)
		echo Aborting
		exit 1
		;;
		*) echo Please enter only y or n
	esac
done

echo Continuing ...

echo ""
echo ""

start=`date +%s`  #Measuring time

#############################################################################################################################


## Generate uniq SraSST Runid based on date hour min sec - Thi ID will be USED in filename
ID=$(date | awk '{print $6$2$3$4}' | sed 's/://g' )

## CreateDirectoryStructure - GlobalResultFile - GlobalGPScoordinates
#QUERY_f="$(basename -- $QUERY)"
QUERY_f="query"
#LIST_f="$(basename -- $LIST)"
LIST_f="list"
RUN="SraSST-"$ID"-"$QUERY_f"_vs_"$LIST_f
echo $RUN
FOLDER=$OUTPUTDIR"/"$RUN
mkdir -p $FOLDER
echo Creating result directory: $FOLDER

PAIRED_READS_f=$FOLDER/"All-retrieved-reads-paired.fastq"; touch $PAIRED_READS_f
UNPAIRED_READS_f=$FOLDER/"All-retrieved-reads-nonpaired.fastq"; touch $UNPAIRED_READS_f
UNPAIRED_READS_fa=$FOLDER/"All-retrieved-reads-nonpaired.fasta"
PAIRED_READS_R1_f=$FOLDER/"All-retrieved-reads-paired_R1.fastq";
PAIRED_READS_R1_fa=$FOLDER/"All-retrieved-reads-paired_R1.fasta";
PAIRED_READS_R2_f=$FOLDER/"All-retrieved-reads-paired_R2.fastq";
PAIRED_READS_R2_fa=$FOLDER/"All-retrieved-reads-paired_R2.fasta";
UNPAIRED_RETRIEVED_READS=$FOLDER/"All-retrieved-reads-nonpaired.besthits.RblastVsOtu.tab"
PAIRED_RETRIEVED_READS_R1=$FOLDER/"All-retrieved-reads-paired_R1.besthits.RblastVsOtu.tab"
PAIRED_RETRIEVED_READS_R2=$FOLDER/"All-retrieved-reads-paired_R2.besthits.RblastVsOtu.tab"

USEARCH_merged_f=$FOLDER/"usearch.merged.fq";
USEARCH_merged_fa=$FOLDER/"usearch.merged.fa";
USEARCH_merged_RETRIEVED=$FOLDER/"usearch.merged.besthits.RblastVsOtu.tab";
USEARCH_filtered_f=$FOLDER/"usearch.filtered.fa";
USEARCH_uniques_f=$FOLDER/"usearch.uniques.fa";
USEARCH_otus_f=$FOLDER/"usearch.otus.fa";
USEARCH_RETRIEVED=$FOLDER/"usearch.otus.NRBlast.tab";
USEARCH_otus_GIDS=$FOLDER/"usearch.otus.NRBlast.gids";
USEARCH_otus_GIDS_NAME=$FOLDER/"usearch.otus.NRBlast.gids-Namelist.txt";

GPS_f=$FOLDER/"GPS_coordinates.tab"; touch $GPS_f
COUNTS_f=$FOLDER/"counts.tab"; touch $COUNTS_f
GLOBAL_f=$FOLDER/"GlobalResults.tab"; touch $GLOBAL_f
LOG_f=$FOLDER/"logs.log"; touch $LOG_f

echo "RUNID"_"LATITUDESTART"_"LONGITUDESTART"_"LATITUDE"_"LONGITUDE"_"LATLON" >> $GPS_f
echo ""QUERY_RUN_Readcount_LATITUDE_LONGITUDE"" | sed 's/_/\t/g' >> $GPS_f
echo Query_READID_PERCID_ALNLGT_MISM_GAP_QSTART_QEND_SbSTART_SbEND_EVAL_BITSCORE_RUN_READ_LONGITUDE_LATITUDE | sed 's/_/\t/g' >> $GLOBAL_f


# Save Original Input for future reference

INPUT_FOLDER=$FOLDER"/input"
mkdir -p $INPUT_FOLDER

cp $QUERY $INPUT_FOLDER"/"
cp $LIST $INPUT_FOLDER"/"


# Fill in run information in .log file
DATE=$(date)
echo $DATE >> $LOG_f
echo "SraSST-"$ID"-"$QUERY"_vs_"$LIST"" >> $LOG_f
echo QUERY is "$QUERY"  fasta queryfile >> $LOG_f
echo LIST is "$LIST"  name of input list of sra RUN  >> $LOG_f
echo PERCID "$PERCID"  blast identity threshold >> $LOG_f
echo MAXTARGET is "$MAXTARGET" "max number of target sequences in blast parameter" >> $LOG_f
echo READCOVER is "$READCOVER"  "percentage read cover for blast algntlgt filtering" >> $LOG_f
echo EVALUE is "$EVALUE"  "Blast evalue parameter" >> $LOG_f
echo "for each $ i sra run (e.g. ERR867907) blastn_vdb command will be" >> $LOG_f
echo "./blastn_vdb -db "$ i" -query $QUERY -outfmt 6 -out ""$QUERY"_vs_"$i".blastres.tab" -perc_identity $PERCID -max_target_seqs $MAXTARGET  -evalue $EVALUE" >> $LOG_f

echo ""
echo ""
echo ""
#
#
#
###Actual start of input processing
for i in $(cat $LIST); 
do
	echo "Processing ""$i"
	INT_FOLDER=$FOLDER/$i
	mkdir $INT_FOLDER
	echo "creating "$INT_FOLDER >> $LOG_f

	##########  BLAST ###############################
	
	echo "Running blastn_vdb"
	BLAST_f=$INT_FOLDER/blastres.tab
	echo ""##blastn_vdb -db "$i" -query "$QUERY" -outfmt 6 -out "$BLAST_f" -perc_identity "$PERCID" -max_target_seqs "$MAXTARGET" -evalue "$EVALUE""" >> $LOG_f
	blastn_vdb -db "$i" -num_threads $NTHREADS -query $QUERY -outfmt 6 -out "$BLAST_f" -perc_identity $PERCID -max_target_seqs $MAXTARGET  -evalue $EVALUE &> $LOG_f

	##########  READCOVER FILTER  ####################

	# checking read length to adapt blast align fiter
	echo "estimating mean read length on first 100 reads"
	ALNLGT=$(fastq-dump --split-files --skip-technical -X 100 -Z $i | grep "length=" | sed 's/ /\t/g'| cut -f3 | sed 's/length=//g' | awk '{ sum += $1; n++ } END { if (n > 0) print sum / n; }'| awk '{print $1 * '"$READCOVER"'}' | cut -f1 -d '.')
	echo "User Setting is" "$READCOVER" ".....SraSST will ONLY select blast alignments over "$ALNLGT"nt" 	
	echo "~"
	echo "~"
	echo "~"
	# Filter results based on blast align filter 
#	#print blastres with aligntlengt filter criteria and extracts hit name
#	(e.g SRA:ERR867907.21038.2) to deduce sra run spot (or read) id number
#	(e.g 21038)
	ALNLGT_f=$INT_FOLDER/"alnlgt_ids"
	cat "$BLAST_f" | awk ' $4 >= '"$ALNLGT"' ' | cut -f2 | tr : '\t' | tr . '\t' | cut -f3 | sort -u > $ALNLGT_f

	RESULT=$(cat "$ALNLGT_f" | wc -l)
	echo "$RESULT" "results to retrieve"

#	############   GPS EXTRACTION  ####################
	if [[ "$RESULT" -gt 0 ]]; then
		LATITUDESTART=$(esearch -db sra -query $i | efetch -format native | xtract -pattern SAMPLE_ATTRIBUTES -block SAMPLE_ATTRIBUTE -if TAG -equals 'Latitude Start' -element VALUE)
		LONGITUDESTART=$(esearch -db sra -query $i | efetch -format native | xtract -pattern SAMPLE_ATTRIBUTES -block SAMPLE_ATTRIBUTE -if TAG -equals 'Longitude Start' -element VALUE)
		LATITUDE=$(esearch -db sra -query $i | efetch -format native | xtract -pattern SAMPLE_ATTRIBUTES -block SAMPLE_ATTRIBUTE -if TAG -equals 'Latitude' -or 'latitude' -element VALUE)
		LONGITUDE=$(esearch -db sra -query $i | efetch -format native | xtract -pattern SAMPLE_ATTRIBUTES -block SAMPLE_ATTRIBUTE -if TAG -equals 'Longitude' -or 'longitude' -element VALUE)
		#the  'geographic location (latitude and longitude)' retrieval is not always working...
		LATLON=$(esearch -db sra -query $i | efetch -format native | xtract -pattern SAMPLE_ATTRIBUTES -block SAMPLE_ATTRIBUTE -if TAG -equals 'lat_lon' -or 'geographic location (latitude and longitude)' -element VALUE)

		echo "$i"_"$LATITUDESTART"_"$LONGITUDESTART"_"$LATITUDE"_"$LONGITUDE"_"$LATLON" >> $GPS_f
		echo "GPS coordinates stored in "$GPS_f
	fi

	##########   READ EXTRACTION   ##################### 
	for k in $(cat "$ALNLGT_f"); do
		FASTQ_f=$INT_FOLDER/$i.$k.fastq-dump
		
		#Fetch retrieved individual fastq read pairs
		fastq-dump --split-files --skip-technical -N $k -X $k -Z $i > $FASTQ_f

		#Add the above to global result fastq file
		READNUMB=$(cat "$FASTQ_f" | grep $i | grep @ | wc -l)
		if [[ "$READNUMB" -eq  2 ]] ; then
			cat $FASTQ_f >> $PAIRED_READS_f
		else 
			cat $FASTQ_f >> $UNPAIRED_READS_f
		fi
		##Append GPS info to blast results to create the Global Read Detail Table
		cat "$BLAST_f" | grep "$i"."$k" |  awk ' {print $1"\t"$2"\t" $3"\t"$4"\t"$5"\t"$6"\t"$7"\t"$8"\t"$9"\t"$10"\t"$11"\t"$12"\t" "'$i'" "\t" "'$k'" "\t" "'$LONGITUDE'" "\t" "'$LATITUDE'"}' >> $GLOBAL_f
		##Disabled here ##### Convert to fasta
		cat $FASTQ_f | awk '{if(NR%4==1) {printf(">%s\n",substr($0,2));} else if(NR%4==2) print;}' > $FASTQ_f.fasta
	done


	### Emptying cache after each run (could be modified to only keep runs matching query)
	echo "Running cache-mgr"
	cache-mgr -c
	
	###Check if file does not contain any fastq file (e.g. no reads were extracted), and delete if true
	echo $INT_FOLDER
	ls "$INT_FOLDER" | grep fastq-dump
	if [[ $? -ne 0 ]]; then
		#rm -r "$INT_FOLDER"
		echo  "$i"" SORRY!! No hits retrieved in this Run " 
		echo  "$i"" SORRY!! No hits retrieved in this Run " >> $LOG_f
		echo "" 
	else                                        
		echo "$i" "HITs found for this Run!!!"
		echo "$i" "HITs found for this Run!!!" >> $LOG_f
		echo "" 
	
		##Feed estimate of results to global result file
		y="$(ls "$INT_FOLDER" | wc -l)"
		## minus 2 to uncount blastid and blastres files --> count indiv reads fastq files
		let "y -= 2"
		echo ""$QUERY"_"$i"_"$y"_"$LATITUDE"_"$LONGITUDE"" | sed 's/_/\t/g' >> $COUNTS_f
	fi
done



###################################################   OTU SEARCH   ###########################################################
#
###Manual creation of global fastq file --> creating R1 and R2 file
#
echo "" 
echo "" 
echo "" 
echo "LAUNCHING USEARCH OTU CALLING (Default parameters accept singletons)"
echo "" 
echo "" 
echo "" 


cat $PAIRED_READS_f | awk 'BEGIN {OFS = "\n"} {header1 = $0 ;
 getline seq1 ; getline qheader1 ; getline qseq1; getline header2; getline seq2; getline qheader2; getline qseq2 ;
  {print header1, seq1, qheader1, qseq1}}' > $PAIRED_READS_R1_f

cat $PAIRED_READS_f | awk 'BEGIN {OFS = "\n"} {header1 = $0 ;
 getline seq1 ; getline qheader1 ; getline qseq1; getline header2; getline seq2; getline qheader2; getline qseq2 ;
  {print header2, seq2, qheader2, qseq2}}' > $PAIRED_READS_R2_f


##Merge read pairs
usearch -fastq_mergepairs $PAIRED_READS_R1_f -fastqout $USEARCH_merged_f

#if "Label Mismatch error" the mentionned reads dont have their respective R1
#or R2 mate, you'll have to delete them before OTU calling"

##Standard usearch filtering parameters
usearch -fastq_filter $USEARCH_merged_f -fastq_maxee 1.0 -fastaout $USEARCH_filtered_f

##Sort uniq reads
usearch -fastx_uniques $USEARCH_filtered_f -relabel Uniq -sizeout -fastaout $USEARCH_uniques_f

##Cluster OTUs  - Option -minsize 1 accepts singleton OTUs - radius 1.5 groups together all reads over 99 id
usearch -cluster_otus $USEARCH_uniques_f -otu_radius_pct 1 -minsize 2 -otus $USEARCH_otus_f -relabel SraSST-"$ID"-Otu

echo "Number of OTUs (including singletons) retrieved by USEARCH for run " $FOLDER  
OTUS_RETRIEVED=$(cat $USEARCH_otus_f | grep '>' | wc -l )
echo $OTUS_RETRIEVED

##Add number of dertemined OTUs to Logfile
echo "Number of OTUs (including singletons) retrieved by USEARCH for run "$RUN >> $LOG_f
echo $OTUS_RETRIEVED >> $LOG_f

##Build custom blastdb based on retrieved OTUs
###########Build future query fasta files
cat $UNPAIRED_READS_f | awk '{if(NR%4==1) {printf(">%s\n",substr($0,2));} else if(NR%4==2) print;}' > $UNPAIRED_READS_fa
cat $PAIRED_READS_R1_f | awk '{if(NR%4==1) {printf(">%s\n",substr($0,2));} else if(NR%4==2) print;}' > $PAIRED_READS_R1_fa
cat $PAIRED_READS_R2_f | awk '{if(NR%4==1) {printf(">%s\n",substr($0,2));} else if(NR%4==2) print;}' > $PAIRED_READS_R2_fa
cat $USEARCH_merged_f | awk '{if(NR%4==1) {printf(">%s\n",substr($0,2));} else if(NR%4==2) print;}' > $USEARCH_merged_fa



# create global blastdb directory in the SraSST workspace, will include the db of differnt runs called by their IDs
DBFOLDER=$FOLDER/SraSSTblastdb
mkdir $DBFOLDER
TITLE=SraSST-"$ID"-otus
BLASTDB_ID=$DBFOLDER/$TITLE

makeblastdb -in $USEARCH_otus_f -dbtype nucl -title $TITLE -out $BLASTDB_ID -parse_seqids


###########Launch blasts with each query file
blastn -num_threads $NTHREADS -db $BLASTDB_ID -query $UNPAIRED_READS_fa -max_target_seqs 1 -outfmt "6" -out $UNPAIRED_RETRIEVED_READS
blastn -num_threads $NTHREADS -db $BLASTDB_ID -query $PAIRED_READS_R1_fa -max_target_seqs 1 -outfmt "6" -out $PAIRED_RETRIEVED_READS_R1
blastn -num_threads $NTHREADS -db $BLASTDB_ID -query $PAIRED_READS_R2_fa -max_target_seqs 1 -outfmt "6" -out $PAIRED_RETRIEVED_READS_R2
blastn -num_threads $NTHREADS -db $BLASTDB_ID -query  $USEARCH_merged_fa -outfmt "6" -max_target_seqs 1 -out $USEARCH_merged_RETRIEVED


##Remote Blast of retrieved OTUs via nr - best hits in nr

echo "Final Remote Blast of retrieved OTUs on NR - This might take a while ... up to 10 minutes"
BLASTDB=/media/data/BLASTDB
#blastn -db nr -query $USEARCH_otus_f -max_target_seqs 1 -outfmt "6" -out $USEARCH_RETRIEVED -remote
blastn -db nr -num_threads $NTHREADS -query $USEARCH_otus_f -max_target_seqs 1 -outfmt "6" -out $USEARCH_RETRIEVED

##Extracting organism names from hits retrieved in NR
cat $USEARCH_RETRIEVED | cut -f2 > $USEARCH_otus_GIDS

for s in $(cat $USEARCH_otus_GIDS); do
 TAXLIN=$(esearch -db nucleotide -query $s | efetch -format xml |xtract -pattern Bioseq -element Textseq-id_accession Seqdesc_title OrgName_lineage Date-std_year | grep "$s")
 echo "$TAXLIN" >> $USEARCH_otus_GIDS_NAME 
 echo "processed "$s""
done


####Warn user about cache size 
end=`date +%s`
runtime=$((end-start))
echo "Moulinette runtime: "$runtime
echo "BEWARE OF CACHE USAGE"
echo "Running cache-mgr ..."
cache-mgr -r
echo "command cache-mgr with option -c to clear you cache - BUT doing so might affect other running instances of SraSST if they share the same ncbi/cache folder."
echo "Your results are stored in "$FOLDER
echo "with the content listed above"
echo "You can find the best NR hits matching your OTUs in " $USEARCH_RETRIEVED
echo "You can find the Organism dESCRIPTION OF THOSE NR hits matching your OTUs in " $USEARCH_otus_GIDS_NAME
echo "You can now MAP your new OTUs to GPS coordinates via reciprocal blast on the retrived reads"
echo "You can also run the tool again using the retrieved unique reads (usearch folder - unique -) to recursively broaden your search"
echo "You're Welcome :-)"

exit 0
